receiver/iisreceiver/config_generator.go

```diff
@@ -0,0 +1,299 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build windows
+// +build windows
+
+package iisreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/iisreceiver"
+
+import (
+	"time"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/iisreceiver/internal/metadata"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/windowsperfcountersreceiver"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.opentelemetry.io/collector/receiver/scraperhelper"
+)
+
+// generateVals Adds dummy data so "Emit" Returns representation of the metrics if they are enabled
+func generateVals(metricBuilder *metadata.MetricsBuilder) {
+	now := pdata.NewTimestampFromTime(time.Now())
+	metricBuilder.RecordIisConnectionActiveDataPoint(now, 0)
+	metricBuilder.RecordIisConnectionAnonymousDataPoint(now, 0)
+	metricBuilder.RecordIisConnectionAttemptCountDataPoint(now, 0)
+	metricBuilder.RecordIisNetworkBlockedDataPoint(now, 0)
+	metricBuilder.RecordIisNetworkFileCountDataPoint(now, 0, "")
+	metricBuilder.RecordIisNetworkIoDataPoint(now, 0, "")
+	metricBuilder.RecordIisRequestCountDataPoint(now, 0, "")
+	metricBuilder.RecordIisRequestQueueAgeMaxDataPoint(now, 0)
+	metricBuilder.RecordIisRequestQueueCountDataPoint(now, 0)
+	metricBuilder.RecordIisRequestRejectedDataPoint(now, 0)
+	metricBuilder.RecordIisThreadActiveDataPoint(now, 0)
+	metricBuilder.RecordIisUptimeDataPoint(now, 0)
+}
+
+// generateWindowsPerfConfig creates a windowssperfreceiver config from dummy metric data and user metricsettings,
+func generateWindowsPerfConfig(scraperCfg scraperhelper.ScraperControllerSettings, metricSettings metadata.MetricsSettings) (*windowsperfcountersreceiver.Config, error) {
+	metricBuilder := metadata.NewMetricsBuilder(metricSettings)
+	generateVals(metricBuilder)
+	metricSlice := metricBuilder.Emit().ResourceMetrics().At(0).InstrumentationLibraryMetrics().At(0).Metrics()
```

Comment:
I think this needs a clear explanation about what it accomplishes and why it's necessary. As I understand it, something like this:

In order to ensure that `metadata.yaml` is the single source of truth for the data model in this receiver, the generated code is reflected upon in and used to generate the config of the `windowsperfcountersreceiver`. `mdatagen` currently does not expose the generated data model directly, so it is instead inferred from an actual result.

Also, because the `windowsperfcountersreceiver` is a generic component, it does not rely on `mdatagen`. Therefore, it does not have a notion of enabled/disabled metrics or any associated filtering. Instead, this receiver must pre-filter those metrics. Like the metric metadata, relying on an actual metric slice allows the config to be the single source of truth for which metrics are collected. Whether each metric is enabled or disable _is_ directly accessible from the `MetricSettings`, but not in a generic way that allows `metadata.yaml` to be the single source of truth. Instead, generating an actual metric slice gives a direct way to collect only the enabled metrics.

This metric slice is used only to generate the config of the `windowsperfcountersreceiver`.


I would also suggest merging the code a bit more to isolate the generation of this dummy slice:
```suggestion
// generateVals Adds dummy data so "Emit" Returns representation of the metrics if they are enabled
func generateVals(metricSettings metadata.MetricsSettings) pdata.MetricSlice {
	metricBuilder := metadata.NewMetricsBuilder(metricSettings)

	now := pdata.NewTimestampFromTime(time.Now())
	metricBuilder.RecordIisConnectionActiveDataPoint(now, 0)
	metricBuilder.RecordIisConnectionAnonymousDataPoint(now, 0)
	metricBuilder.RecordIisConnectionAttemptCountDataPoint(now, 0)
	metricBuilder.RecordIisNetworkBlockedDataPoint(now, 0)
	metricBuilder.RecordIisNetworkFileCountDataPoint(now, 0, "")
	metricBuilder.RecordIisNetworkIoDataPoint(now, 0, "")
	metricBuilder.RecordIisRequestCountDataPoint(now, 0, "")
	metricBuilder.RecordIisRequestQueueAgeMaxDataPoint(now, 0)
	metricBuilder.RecordIisRequestQueueCountDataPoint(now, 0)
	metricBuilder.RecordIisRequestRejectedDataPoint(now, 0)
	metricBuilder.RecordIisThreadActiveDataPoint(now, 0)
	metricBuilder.RecordIisUptimeDataPoint(now, 0)
	
	return metricBuilder.Emit().ResourceMetrics().At(0).InstrumentationLibraryMetrics().At(0).Metrics()
}

// generateWindowsPerfConfig creates a windowssperfreceiver config from dummy metric data and user metricsettings,
func generateWindowsPerfConfig(scraperCfg scraperhelper.ScraperControllerSettings, metricSettings metadata.MetricsSettings) (*windowsperfcountersreceiver.Config, error) {
	metricSlice := generateDummyMetrics(metricSettings)
```

Reply:
Safe to say there are some opportunities to propose enhancements to `mdatagen` here. I'll give this some more thought, but generally speaking, I think this demonstrates that `mdatagen` should 1) expose metric _metadata_ directly, and 2) provide a way to consume metric settings in a generic way. Maybe something like:

```go
for _, m := metricBuilder.EnabledMetrics(metricSettings) {
    innerConfig.metrics = append(innerConfig.metrics, toConfigItem(m))
}

func toConfigItem(m metadata.MetricDefinition) *PerfCounterMetricConfig {
    ... reflect on abstract representation of metric to produce request item
}

```

Reply:
Updated